From 9ecdcc95bbb843e5936fddc03a1077c83ede03fc Mon Sep 17 00:00:00 2001
From: Yoni Farhadian <yonif@marvell.com>
Date: Tue, 30 Jul 2013 13:33:31 +0300
Subject: [PATCH 0907/1825] alp: PPv2: fix: Switch MTU change

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit 1c6f794a40d2e496f56a2f5ef34f33f73f35fc61

	- Fix change switch's MTU when changing master interface's MTU
	- Handle case of changing MTU in switch transparent preset mode
	- Update switch's MTU when activating master interface

Change-Id: I880acfc16c9952630323017533cb594aabc1d184
Signed-off-by: Yoni Farhadian <yonif@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/2883
Reviewed-by: Uri Eliyahu <uriel@marvell.com>
Reviewed-by: Igor Patrik <igorp@marvell.com>
Tested-by: Igor Patrik <igorp@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 .../mv_drivers_lsp/mv_mux/mv_mux_netdev.c          |   43 ++++++++++++--------
 .../mv_drivers_lsp/mv_mux/mv_mux_netdev.h          |    1 +
 2 files changed, 27 insertions(+), 17 deletions(-)

diff --git a/arch/arm/plat-armada/mv_drivers_lsp/mv_mux/mv_mux_netdev.c b/arch/arm/plat-armada/mv_drivers_lsp/mv_mux/mv_mux_netdev.c
index cbd7563..1cccb41 100644
--- a/arch/arm/plat-armada/mv_drivers_lsp/mv_mux/mv_mux_netdev.c
+++ b/arch/arm/plat-armada/mv_drivers_lsp/mv_mux/mv_mux_netdev.c
@@ -172,6 +172,8 @@ void mv_mux_switch_attach(int gbe_port, int preset, int vid, int tag, int switch
 	mux_switch_shadow.switch_port = switch_port;
 	mux_switch_shadow.gbe_port = gbe_port;
 	mux_switch_shadow.attach = MV_TRUE;
+	/* Update MTU when activating master interface */
+	mux_switch_shadow.mtu = -1;
 
 #ifdef CONFIG_MV_INCLUDE_SWITCH
 	/* set mux ops to be used by switch driver */
@@ -443,14 +445,12 @@ int mv_mux_open(struct net_device *dev)
 	netif_stacked_transfer_operstate(root, dev);
 	netif_tx_wake_all_queues(dev);
 
+	if (dev->mtu > root->mtu)
+		dev->mtu = root->mtu;
+
 #ifdef CONFIG_MV_INCLUDE_SWITCH
-	if (mux_switch_shadow.attach) {
-		if (dev->mtu != root->mtu) {
-			dev->mtu = root->mtu;
-			mv_mux_switch_mtu_update(dev->mtu);
-		}
+	if (mux_switch_shadow.attach)
 		mv_switch_group_enable(pmux_priv->idx);
-	}
 #endif
 
 	printk(KERN_NOTICE "%s: started\n", dev->name);
@@ -621,11 +621,6 @@ static struct net_device *mv_mux_netdev_init(int port, struct net_device *mux_de
 
 	SET_ETHTOOL_OPS(mux_dev, &mv_mux_tool_ops);
 
-#ifdef CONFIG_MV_INCLUDE_SWITCH
-	if (mux_switch_shadow.attach)
-		mv_mux_switch_mtu_update(mux_dev->mtu);
-#endif
-
 	return mux_dev;
 }
 /*-----------------------------------------------------------------------------------------*/
@@ -817,7 +812,11 @@ static int mux_device_event(struct notifier_block *unused, unsigned long event,
 
 	/* exit - if not mux device */
 	if (tag_type == MV_TAG_TYPE_NONE)
-		goto out;
+#ifdef CONFIG_MV_INCLUDE_SWITCH
+		/* Check that that this port is not CPU port (can be untagged) */
+		if (mux_switch_shadow.attach && (mux_switch_shadow.gbe_port != port))
+#endif
+			goto out;
 
 	switch (event) {
 
@@ -862,13 +861,14 @@ static int mux_device_event(struct notifier_block *unused, unsigned long event,
 		while (mux_dev != NULL) {
 			pdev_priv = MV_MUX_PRIV(mux_dev);
 			dev_set_mtu(mux_dev, dev->mtu);
-#ifdef CONFIG_MV_INCLUDE_SWITCH
-			if (mux_switch_shadow.attach)
-				mv_mux_switch_mtu_update(mux_dev->mtu);
-#endif
 			mux_dev = pdev_priv->next;
 		}
-
+#ifdef CONFIG_MV_INCLUDE_SWITCH
+		if (mux_switch_shadow.attach) {
+			mux_switch_shadow.mtu = dev->mtu;
+			mv_mux_switch_mtu_update(dev->mtu);
+		}
+#endif
 		break;
 
 	case NETDEV_DOWN:
@@ -903,6 +903,15 @@ static int mux_device_event(struct notifier_block *unused, unsigned long event,
 			/* dev_change_flags call to mv_mux_open*/
 			mux_dev = pdev_priv->next;
 		}
+
+#ifdef CONFIG_MV_INCLUDE_SWITCH
+		/* Check for MTU updates */
+		if (mux_switch_shadow.attach &&
+				((mux_switch_shadow.mtu == -1) || (mux_switch_shadow.mtu > dev->mtu))) {
+			mux_switch_shadow.mtu = dev->mtu;
+			mv_mux_switch_mtu_update(dev->mtu);
+		}
+#endif
 		break;
 
 	case NETDEV_FEAT_CHANGE:
diff --git a/arch/arm/plat-armada/mv_drivers_lsp/mv_mux/mv_mux_netdev.h b/arch/arm/plat-armada/mv_drivers_lsp/mv_mux/mv_mux_netdev.h
index f647b7a..0c6ad43 100644
--- a/arch/arm/plat-armada/mv_drivers_lsp/mv_mux/mv_mux_netdev.h
+++ b/arch/arm/plat-armada/mv_drivers_lsp/mv_mux/mv_mux_netdev.h
@@ -73,6 +73,7 @@ struct mv_mux_switch_port {
 	int    vid;
 	int    switch_port;
 	int    gbe_port;
+	int    mtu;
 	bool   attach;
 };
 
-- 
1.7.5.4

